group-coordinator/src/test/java/org/apache/kafka/coordinator/group/RecordSerdeTest.java

```diff
@@ -0,0 +1,260 @@
+            new ConsumerGroupMetadataValue().setEpoch(10),
+            (short) 0
+        );
+        ByteBuffer valueBuffer = MessageUtil.toVersionPrefixedByteBuffer(value.version(), value.message());
+
+        Record record = serDe.deserialize(keyBuffer, valueBuffer);
+        assertEquals(key, record.key());
+        assertEquals(value, record.value());
+    }
+
+    @Test
+    public void testDeserializeWithTombstoneForValue() {
+        RecordSerde serDe = new RecordSerde();
+
+        ApiMessageAndVersion key = new ApiMessageAndVersion(
+            new ConsumerGroupMetadataKey().setGroupId("foo"),
+            (short) 3
+        );
+        ByteBuffer keyBuffer = MessageUtil.toVersionPrefixedByteBuffer(key.version(), key.message());
+
+        Record record = serDe.deserialize(keyBuffer, null);
+        assertEquals(key, record.key());
+        assertNull(record.value());
+    }
+
+    @Test
+    public void testDeserializeWithInvalidRecordType() {
+        RecordSerde serDe = new RecordSerde();
+
+        ByteBuffer keyBuffer = ByteBuffer.allocate(64);
+        keyBuffer.putShort((short) 255);
+        keyBuffer.rewind();
+
+        ByteBuffer valueBuffer = ByteBuffer.allocate(64);
+
+        CoordinatorLoader.UnknownRecordTypeException ex =
+            assertThrows(CoordinatorLoader.UnknownRecordTypeException.class,
+                () -> serDe.deserialize(keyBuffer, valueBuffer));
+        assertEquals((short) 255, ex.unknownType());
+    }
+
+    @Test
+    public void testDeserializeWithKeyEmptyBuffer() {
+        RecordSerde serDe = new RecordSerde();
+
+        ByteBuffer keyBuffer = ByteBuffer.allocate(0);
+        ByteBuffer valueBuffer = ByteBuffer.allocate(64);
+
+        RuntimeException ex =
+            assertThrows(RuntimeException.class,
+                () -> serDe.deserialize(keyBuffer, valueBuffer));
+        assertEquals("Could not read version from key's buffer.", ex.getMessage());
+    }
+
+    @Test
+    public void testDeserializeWithValueEmptyBuffer() {
+        RecordSerde serDe = new RecordSerde();
+
+        ApiMessageAndVersion key = new ApiMessageAndVersion(
+            new ConsumerGroupMetadataKey().setGroupId("foo"),
+            (short) 3
+        );
+        ByteBuffer keyBuffer = MessageUtil.toVersionPrefixedByteBuffer(key.version(), key.message());
+
+        ByteBuffer valueBuffer = ByteBuffer.allocate(0);
+
+        RuntimeException ex =
+            assertThrows(RuntimeException.class,
+                () -> serDe.deserialize(keyBuffer, valueBuffer));
+        assertEquals("Could not read version from value's buffer.", ex.getMessage());
+    }
+
+    @Test
+    public void testDeserializeWithInvalidKeyBytes() {
+        RecordSerde serDe = new RecordSerde();
+
+        ByteBuffer keyBuffer = ByteBuffer.allocate(2);
+        keyBuffer.putShort((short) 3);
+        keyBuffer.rewind();
+
+        ByteBuffer valueBuffer = ByteBuffer.allocate(2);
```

Comment:
To be clear, this is also invalid, but we read the key first so we don't hit it?